compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -1018,8 +1029,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         debug!("coercion::can_with_predicates({:?} -> {:?})", source, target);
 
         let cause = self.cause(DUMMY_SP, ObligationCauseCode::ExprAssignable);
-        // We don't ever need two-phase here since we throw out the result of the coercion
-        let coerce = Coerce::new(self, cause, AllowTwoPhase::No);
+        // We don't ever need two-phase here since we throw out the result of the coercion.
+        // We also just always set `coerce_never` to true, since this is a heuristic.
+        let coerce = Coerce::new(self, cause, AllowTwoPhase::No, true);
```

Comment:
hmm, that means we can end up with `can_coerce` returning false and then coercing failing. I'd have to look at how this is used to check whether it may be a problem (or you tell me)

Reply:
Well firstly, `can_coerce` is called on arbitrary types that are not associated with values or places, so getting this 100% right is literally intractible.

`can_coerce` is used like `can_eq`, where it is *heuristic*. It should never be used for *correctness*, since it also records no adjustments and has no side-effects.

Reply:
yes, I wanted to know whether we use it anywhere to decide "yes, this is a coercion" only to later error. i.e. whether this results in unintended errors

Reply:
Yeah, I checked and there is none. Only in suggestion code or clippy.

Reply:
Well ideally it only returns false negatives since we introduce coercions when it returns true. 

But right now it's only more correct than the status quo, and not 100% sound. e.g. in cases where matches may or may not read, like `(_ | 1)` in a pattern.

Reply:
That pattern never reads, right? In left-to-right order, `_` always matches, so we never reach the `1`.

Reply:
right, but to detect such cases we need to consider arbitrarily complex patterns.

Reply:
So what is the consequence of not doing that? You said it's unsound -- could you give an example?

Reply:
> You said it's unsound

Because the pattern does not read but we consider it to read because one of the subpatterns reads, inserting a `NeverToAny` coercion is unsound.

Reply:
Right but what does that mean? In terms of the surface language? That's why I asked for an example. :)

Following the other threads here, it seems there's actually quite a bit of design space. When this is done, could you write a spec of what got implemented -- i.e., which patterns in `let pat = *ptr` actually cause a value to be manifest?

FWIW I would expect `let (_, _) = *ptr;` on a pair to be equivalent to `let _`, so I am a bit surprised when you said that `Struct { .. }` is considered to produce a value. In fact things can get more subtle than value vs no value -- with enums, it's really about which operations do we actually perform on them? In `match`, we might do a discriminant read but then not read any of the fields; the same could presumably happen here?

Reply:
Frankly I do not have the energy or want to put in any effort to fix this any further.

Reply:
Like, I literally don't care about the cases which are up for debate or which may or may not produce a read depending on the MIR, because these checks are performed on the HIR and I do not have the time or energy to make these checks more complicated than they are now. I am setting out in this PR to fix the egregiously unsound cases ONLY, and improving these semantics further can be done as a follow-up.

Reply:
I don't even know which cases you're talking about, is the problem.^^ I don't know HIR. So I don't even know what the questions are, which cases you consider "egregiously unsound", and which "not-so-egregiously unsound". That's why I asked, once this PR lands, for an explanation of what even got implemented.

Reply:
This PR implements a heuristic which disables two things that are currently being performed on the HIR when we have expressions that involve place-like expressions that point to `!`. Specifically, it will:

1. Disable the `NeverToAny` coercion we implicitly insert.
2. Disable the logic which considers the expression to diverge, which affects the type returned by the containing block.

We do those IF the expression is a place-like expression, and the parent expression is either:
1. the LHS of an assignment
2. AddrOf
3. A match or let IF all of the *patterns consitute a read*

And finally, a pattern currently is considered to constitute a read UNLESS it is a wildcard or a OR of wildcards. All other patterns are considered currently to constitute a read. Specifically, because `NeverToAny` is a coercion performed on a *value* and not a *place*, `Struct { .. }` on a `!` type must be a coercion currently. This is already considered UB by [miri](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=daf3a2246433fe43fdc07d1389c276c9).


That means it does not affect the preexisting UB in this case:

```
let Struct { .. } = *never_ptr;
```

Even though it's likely up for debate, it almost certainly causes inference changes which I do *NOT* want to fix in this PR.

Reply:
The "not so egregiously unsound" case is basically the last example I gave, since that is arguably *not* even unsound but just regular UB because users already need to accept that they're doing sketchy stuff by wrapping their code in `unsafe {}`, and the fact that the coercion there is implicitly occurring is kinda just 🤷.

Recall that I'm primarily motivated by trying to make `&raw` sound, and none of this affects that. The only stuff that's "up for debate" here is just the semantics surrounding pattern matching, which I think are probably incredibly rare, already *more* UB than they are than after this PR, and are easily refined later.

Reply:
I do want to apologize though, my earlier comment here was a bit dramatic since I have (not the fault of the reviewers, of course) been a bit overwhelmed by the process of carrying out this work with fixing never+place exprs in the HIR, especially since my original desire was just to fix the simplest cases of this (e.g. the AddrOf case).

I do agree that there is a lot of design space for how we treat patterns, and I wasn't clear that when I said I didn't really want to further work on this, I really meant I didn't want to tweak this PR in cases that are presumably more "up for debate" like struct patterns that do not read, especially given how the compiler treats them on nightly. 

I can certainly document where this PR ends up if and when it does actually land, though I am not too certain where that documentation would actually live.

Reply:
> But here *ptr does not have type ! and IIUC the special behavior that triggers the coercion is specific to !, it does not apply to all uninhabited types?

Correct. We only perform this behavior of `!` and no other uninhabited types. That's partly why I renamed this function to `..._guaranteed_to_constitute_read_for_never`, since it only really is tuned for `!`.